piet-common/src/cairo_back.rs

```diff
@@ -101,33 +101,60 @@ impl<'a> BitmapTarget<'a> {
         CairoRenderContext::new(&mut self.cr)
     }
 
-    /// Get raw RGBA pixels from the bitmap.
-    pub fn into_raw_pixels(mut self, fmt: ImageFormat) -> Result<Vec<u8>, piet::Error> {
+    /// Get raw RGBA pixels from the bitmap by copying them into `buf`. If all the pixels were
+    /// copied, returns the number of bytes written. If `buf` wasn't big enough, returns an error
+    /// and doesn't write anything.
+    pub fn copy_raw_pixels(
+        &mut self,
+        fmt: ImageFormat,
+        buf: &mut [u8],
+    ) -> Result<usize, piet::Error> {
         // TODO: convert other formats.
         if fmt != ImageFormat::RgbaPremul {
             return Err(piet::Error::NotSupported);
         }
-        std::mem::drop(self.cr);
         self.surface.flush();
         let stride = self.surface.get_stride() as usize;
         let width = self.surface.get_width() as usize;
         let height = self.surface.get_height() as usize;
-        let mut raw_data = vec![0; width * height * 4];
-        let buf = self
-            .surface
-            .get_data()
-            .map_err(Into::<Box<dyn std::error::Error>>::into)?;
-        for y in 0..height {
-            let src_off = y * stride;
-            let dst_off = y * width * 4;
-            for x in 0..width {
-                raw_data[dst_off + x * 4 + 0] = buf[src_off + x * 4 + 2];
-                raw_data[dst_off + x * 4 + 1] = buf[src_off + x * 4 + 1];
-                raw_data[dst_off + x * 4 + 2] = buf[src_off + x * 4 + 0];
-                raw_data[dst_off + x * 4 + 3] = buf[src_off + x * 4 + 3];
-            }
+        let size = width * height * 4;
+        if buf.len() < size {
+            return Err(piet::Error::InvalidInput);
         }
-        Ok(raw_data)
+        unsafe {
+            // Cairo's rust wrapper has extra safety checks that we want to avoid: it won't let us
+            // get the data from an ImageSurface that's still referenced by a context. The C docs
+            // don't seem to think that's a problem, as long as we call flush (which we already
+            // did), and promise not to mutate anything.
+            let data = {
+                let data_ptr = cairo_sys::cairo_image_surface_get_data(self.surface.to_raw_none());
+                if data_ptr.is_null() {
+                    let err = cairo::BorrowError::from(cairo::Status::SurfaceFinished);
+                    return Err((Box::new(err) as Box<dyn std::error::Error>).into());
+                }
+                std::slice::from_raw_parts(data_ptr, size)
```

Comment:
I'm not sure that using `size` here is correct; what happens if `stride` is different from `width`?

Reply:
Also (and please forgive my spam) one other little thing: I would try to figure out what the smallest `unsafe` function we can write is, and write that (the byte copying stuff below doesn't need to be unsafe, for instance.)

Reply:
If `stride` is different from `width`, we potentially read a different region of the source data. But no matter what, the part of `buf` that gets written is `[0..size]`.

Regarding the unsafe block, rustc will accept
````rust
    let data = unsafe { ... };
```

I made it bigger, so that it encompasses the entire region where we need to maintain safety-related invariants. (Specifically, we need to avoid any mutations while `data` is live.) Do you still prefer the smaller version?

Reply:
but we're creating a slice with a size that doesn't take the `stride` into account, so that we can potentially index out of bounds? This isn't a soundness concern (we won't ever be creating a slice that is larger than our source buffer) but we might at least crash on a bounds check?

Also thinking about that it's possible that this is bounds checking every index, which would be annoying, and might be something to check if we're worried about performance.

Reply:
Ah, sorry, I thought you were talking about the returned size value :facepalm: I'll fix this...

Reply:
Regarding the bounds checks, unfortunately you're right. Here's the inner loop, with 8 bounds checks per iteration.
```
.LBB12_7:
 leaq    (%r10, %rcx), %rsi
 addq    $2, %rsi
 cmpq    %rdx, %rsi
 jae     .LBB12_24
 leaq    (%rcx, %rbp), %rsi
 cmpq    %r15, %rsi
 jae     .LBB12_25
 movzbl  -1(%rax, %rcx), %ebx
 movb    %bl, -3(%r14, %rcx)
 leaq    (%r10, %rcx), %rsi
 addq    $1, %rsi
 cmpq    %rdx, %rsi
 jae     .LBB12_27
 leaq    (%rcx, %rbp), %rsi
 addq    $1, %rsi
 cmpq    %r15, %rsi
 jae     .LBB12_28
 leaq    (%r10, %rcx), %rsi
 movzbl  -2(%rax, %rcx), %ebx
 movb    %bl, -2(%r14, %rcx)
 cmpq    %rdx, %rsi
 jae     .LBB12_29
 leaq    (%rcx, %rbp), %rsi
 addq    $2, %rsi
 cmpq    %r15, %rsi
 jae     .LBB12_30
 movzbl  -3(%rax, %rcx), %ebx
 movb    %bl, -1(%r14, %rcx)
 leaq    (%r10, %rcx), %rsi
 addq    $3, %rsi
 cmpq    %rdx, %rsi
 jae     .LBB12_31
 leaq    (%rcx, %rbp), %rsi
 addq    $3, %rsi
 cmpq    %r15, %rsi
 jae     .LBB12_32
 addq    $1, %rdi
 movzbl  (%rax, %rcx), %ebx
 movb    %bl, (%r14, %rcx)
 addq    $4, %rcx
 cmpq    %r12, %rdi
 jb      .LBB12_7
```

After adding get_unchecked:
```
 .LBB12_20:
 movdqu  xmm3, xmmword, ptr, [rcx, +, 4*rbx]
 movdqu  xmm2, xmmword, ptr, [rcx, +, 4*rbx, +, 16]
 movdqu  xmm1, xmmword, ptr, [rcx, +, 4*rbx, +, 32]
 movdqu  xmm4, xmmword, ptr, [rcx, +, 4*rbx, +, 48]
 movdqa  xmm5, xmm2
 punpckhbw xmm5, xmm0
<snip>
 packuswb xmm1, xmm5
 movdqu  xmmword, ptr, [rsi, +, 4*rbx, +, 32], xmm1
 movdqu  xmmword, ptr, [rsi, +, 4*rbx, +, 48], xmm4
 movdqu  xmmword, ptr, [rsi, +, 4*rbx], xmm3
 movdqu  xmmword, ptr, [rsi, +, 4*rbx, +, 16], xmm2
 add     rbx, 16
 cmp     r11, rbx
 jne     .LBB12_20
```
Unrolling and autovectorization ftw!